### DISCUSSION TITLE
`eglot--sig-info` misbehavior for ocaml

### DISCUSSION THREAD

**Discussion by jfeser** [joaotavora/eglot#1251] [Q&A] [answered]
`eglot--sig-info` tries to highlight functions in signature help; unfortunately, this interacts badly with ocaml syntax. Could this code be made configurable or removed?

Relevant code
```elisp
      ;; Ad-hoc attempt to parse label as <name>(<params>)
        (when (looking-at "\\([^(]*\\)(\\([^)]+\\))")
          (add-face-text-property (match-beginning 1) (match-end 1)
                                  'font-lock-function-name-face))
```

Behavior:
<img width="853" alt="Screenshot 2023-07-01 at 4 17 26 PM" src="https://github.com/joaotavora/eglot/assets/1149464/0b43d0c8-b63e-4ff9-972d-00256a4fde4c">

**Comment by joaotavora** [comment]
I don't know ocaml, though I have a t-shirt :)

So I need you to point out exactly what the undesired behavior is and what the desired behavior would be. Use plain text, if possible, though a localized screenshot with a comment explaining exactly what I'm supposed to look at is also preferable.

As it stands, I don't know exactly what is wrong.

**Reply by jfeser** [reply]
Wow, what a quick response!

The part in the box is the function name, so it's the only part that should have `font-lock-function-name-face`. The other highlighted parts are part of the parameter list: 
<img width="791" alt="Screenshot 2023-07-01 at 4 17 26 PM" src="https://github.com/joaotavora/eglot/assets/1149464/d4793abe-e088-4775-9cb9-c2249abfce5a">

I'd be happy to open a pr to configure the highlighting heuristic if that's desired.

**Reply by joaotavora** [reply]
Ok, but you didn't understand yet what in asked for. I need a before and after, an "observed" and "expected".To understand the problem. Please. Then we can discuss a solution.

**Reply by jfeser** [reply]
Ok, here's an "expected" version:

<img width="853" alt="Screenshot 2023-07-01 at 4 42 54 PM" src="https://github.com/joaotavora/eglot/assets/1149464/93fe159e-c963-47fd-b5f6-64217f1ed36a">

**Reply by joaotavora** [reply]
Alright, thanks. But no purple highlighting whatsoever? 

Also, can you find the bit of the Eglot events buffer that corresponds to this signature report and paste it here as text, please?

**Reply by jfeser** [reply]
Highlighting just `Document.Merlin.with_pipeline_exn` would be ideal, but no highlighting is better than too much.

Events:
```
[client-request] (id:297) Sat Jul  1 16:55:06 2023:
(:jsonrpc "2.0" :id 297 :method "textDocument/signatureHelp" :params
          (:textDocument
           (:uri "file:///Users/feser/work/ocaml-workspace/ocaml-lsp/ocaml-lsp-server/src/signature_help.ml")
           :position
           (:line 199 :character 35)))
[server-reply] (id:296) Sat Jul  1 16:55:06 2023:
(:id 296 :jsonrpc "2.0" :result
     (:contents
      (:kind "markdown" :value "```ocaml\n?name:string ->\nDocument.Merlin.t ->\n(Mpipeline.t -> 'a) ->\n'a Fiber.t\n```")
      :range
      (:end
       (:character 39 :line 199)
       :start
       (:character 6 :line 199))))
[server-reply] (id:297) Sat Jul  1 16:55:06 2023:
(:id 297 :jsonrpc "2.0" :result
     (:activeParameter 1 :activeSignature 0 :signatures
                       [(:label "Document.Merlin.with_pipeline_exn : ?name:string -> Document.Merlin.t -> (Mpipeline.t -> application_signature option) -> application_signature option Fiber.t" :parameters
                                [(:label
                                  [36 48])
                                 (:label
                                  [52 69])
                                 (:label
                                  [73 118])])]))
```

**Reply by joaotavora** [reply]
> Highlighting just Document.Merlin.with_pipeline_exn would be ideal, but no highlighting is better than too much.

Agree. 

This function is hairy enough, but I guess one could highlight only until the start position of the first parameter, when that is available, instead of doing that silly regexp.

A customization point is overkill here: this is too tiny a detail though.

**Reply by joaotavora** [reply]
Can you try this patch?
```diff
diff --git a/lisp/progmodes/eglot.el b/lisp/progmodes/eglot.el
index e2478f2dde3..94ca5c94245 100644
--- a/lisp/progmodes/eglot.el
+++ b/lisp/progmodes/eglot.el
@@ -3196,11 +3196,8 @@ eglot--sig-info
                  ((:documentation sigdoc)) parameters activeParameter)
       sig
     (with-temp-buffer
-      (save-excursion (insert siglabel))
-      ;; Ad-hoc attempt to parse label as <name>(<params>)
-        (when (looking-at "\\([^(]*\\)(\\([^)]+\\))")
-          (add-face-text-property (match-beginning 1) (match-end 1)
-                                  'font-lock-function-name-face))
+        (insert siglabel)
+        ;; Ad-hoc attempt to parse label as <name>(<params>)
         ;; Add documentation, indented so we can distinguish multiple signatures
         (when-let (doc (and (not briefp) sigdoc (eglot--format-markup sigdoc)))
           (goto-char (point-max))
@@ -3213,8 +3210,13 @@ eglot--sig-info
                         ((:label parlabel))
                         ((:documentation pardoc)))
              parameter
+           (when (zerop i)
+             (goto-char (elt parlabel 0))
+             (search-backward "(" nil t)
+             (add-face-text-property (point-min) (point)
+                                     'font-lock-function-name-face))
            ;; ...perhaps highlight it in the formals list
-           (when (and (eq i active-param))
+           (when (= i active-param)
              (save-excursion
                (goto-char (point-min))
                (pcase-let
```

**Reply by jfeser** [reply]
That's a significant improvement. Thanks for the quick fix!

**Reply by joaotavora** [reply]
Thanks for testing. I've already pushed it to Emacs master, so it will be in the next Eglot 1.16 release (but you can get it from https://elpa.gnu.org/devel/eglot.html sooner than that, probably).

When you say "it's an improvement", do you think anything is still missing?  Can you show me a screenshot of the current behavior with the patch?

**Reply by jfeser** [reply]
<img width="853" alt="Screenshot 2023-07-01 at 8 44 55 PM" src="https://github.com/joaotavora/eglot/assets/1149464/515dddde-1937-428f-814d-ee57b1402783">

Technically, the ':' isn't part of the function name, so shouldn't be highlighted, but I think it's a only a minor issue.

**Reply by joaotavora** [reply, marked as answer]
Ahh, yes.  I've pushed a tweak which seems more correct and general.
```diff
commit c2e4c68333d72a342bb531449412a424b3507a65
Author: João Távora <joaotavora@gmail.com>
Date:   Sun Jul 2 11:57:15 2023 +0100

    Eglot: another tweak to eglot--sig-info
    
    Going to the start of the first param and skipping non-word syntax
    backward would seem to be the right thing to do, for traditional
    C-style languages and also languages like Ocaml where function
    signatures don't start the param list with '('.
    
    See also https://github.com/joaotavora/eglot/discussions/1251.
    
    * lisp/progmodes/eglot.el (eglot--sig-info): Rework again.

diff --git a/lisp/progmodes/eglot.el b/lisp/progmodes/eglot.el
index 897cf3bc93e..f09c348143d 100644
--- a/lisp/progmodes/eglot.el
+++ b/lisp/progmodes/eglot.el
@@ -3212,7 +3212,7 @@ eglot--sig-info
            parameter
          (when (zerop i)
            (goto-char (elt parlabel 0))
-           (search-backward "(" nil t)
+           (skip-syntax-backward "^w")
            (add-face-text-property (point-min) (point)
                                    'font-lock-function-name-face))
          ;; ...perhaps highlight it in the formals list
```
